Update profile information

Eeyore6 opened the pull request:
![image](https://github.com/mursal-ashraf/bankan/assets/100513480/9ac5740c-4f68-424e-8963-45de32adcd13)
The information saved will be remain whenever you log in. If you delete the information, it will not come back until you re-type them.

Eeyore6 committed:
Update Profile.tsx

Made the Email Uneditable

Eeyore6 committed:
Update Profile.tsx

Store the information to raw_user_meta_data. Used a handleSave funciton, consider to make it a common save function into database.

Eeyore6 commented:
Will remove consolelog part later

Eeyore6 committed:
Update the footer layout

A small adjustment for the footer layout

Eeyore6 commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Create user data on supabase on user signup. (under auth->users -> raw_user_meta)](https://trello.com/c/i4ytCX1G/41-create-user-data-on-supabase-on-user-signup-under-auth-users-rawusermeta)

Eeyore6 committed:
Update Profile.tsx

Make an edit state (information can be edited when the edit button is clicked, otherwise not editable)

Eeyore6 commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Create edit state for profile page.](https://trello.com/c/isiZefVa/44-create-edit-state-for-profile-page)

mursal-ashraf commented on frontend/src/components/Profile/Profile.tsx:
```diff
@@ -32,122 +48,155 @@ const InnerProfile: React.FC = () => {
+            variant="filled"
+            fullWidth
+            InputProps={{ style: { fontSize: 20, height: '3rem' } }}
+            disabled
```
What's the reason for having email permanently disabled?

Eeyore6 replied:
Emial is not editable cuz we want the email as the logging in's account

mursal-ashraf replied:
Then we can just remove the field and not show it, I reckon that's better (only when we're editing).

Eeyore6 replied:
if we remove the field, the page will be empty if not click edit. We can keep the field to show the details or as a entry for user to input.

mursal-ashraf commented on frontend/src/components/Profile/Profile.tsx:
```diff
@@ -32,122 +48,155 @@ const InnerProfile: React.FC = () => {
+        console.log("User updated:", data);
+      }
+    }
+    setIsEditing(false);
```
I think we should only `setIsEditing` to false when the save is successful right? If the save fails we should leave it in the editing state

Eeyore6 replied:
yeah, we should include the save fails situation.

mursal-ashraf commented on frontend/src/components/Profile/Profile.tsx:
```diff
@@ -9,19 +9,35 @@ import {
 
   const [open, setOpen] = useState(false);
   const [image, setImage] = useState<string | null>(null);
+  const [formData, setFormData] = useState<UserMetadata>(defaultFormData);
```
Instead of `defaultFormData`, we can put `user.metadata` right?

mursal-ashraf replied:
In fact, you can just put `{ ...user.meta_data, email: user.email as string }` right?

mursal-ashraf replied:
That way we can get rid of the useEffect on line 51. And less code === better code. And we can also get rid of the `defaultFormData` then as well

Eeyore6 replied:
I have the defaultFormData for the unlogged in state,cuz as we logged out, it wont turn to Home page but stay the profile. So when the user is logged out, it will show empty state.

mursal-ashraf replied:
Ah yes. I think you can remove it still, we will do another card/PR for making sure that if a page requires a user to be logged in, they will be required to login first, then go to the page.

Eeyore6 replied:
Yeah, we can separate the Check user logged for each page. Will remove the default form data

Eeyore6 replied:
![image](https://github.com/mursal-ashraf/bankan/assets/100513480/b823a35d-644f-4164-b87d-61dd0682e900)
Edit the default Form Date

mursal-ashraf commented on frontend/src/components/Profile/Profile.tsx:
```diff
@@ -32,122 +48,155 @@ const InnerProfile: React.FC = () => {
+  }, [user]);
+
+  const handleSave = async () => {
+    if (user) {
```
What's the reason for this user check here? It's not being used in the rest of the function.

Eeyore6 replied:
The if statement can be removed since we have default form for unlogged in state

mursal-ashraf commented on frontend/src/components/Profile/Profile.tsx:
```diff
@@ -32,122 +48,155 @@ const InnerProfile: React.FC = () => {
+            className="text-xl mb-6"
+            style={{ padding: '8px 24px' }}
+            onClick={() => setOpen(true)}
+            disabled={!isEditing}
```
Is this diabled necessary? It wouldn't show up if isEditing was false right?

Eeyore6 replied:
Upload button will show up when edit is clicked, and you are right, since it wont show up, user cannot click it. Will remove the disable attribute

mursal-ashraf commented on frontend/src/components/Profile/Profile.tsx:
```diff
@@ -9,19 +9,35 @@ import {
 
 const InnerProfile: React.FC = () => {
-  const [formData, setFormData] = useState({
+  type UserMetadata = {
```
Can this be put in a separate file please. Helps keep things neat and tidy (and reusable)

Eeyore6 replied:
Will put innerProfile into signle file. Thanks Mursal, it is a good suggestion.

Eeyore6 replied:
And I have tried changing and updating details, the supabase is sucessfully updated

mursal-ashraf replied:
> Will put innerProfile into signle file. Thanks Mursal, it is a good suggestion.

Sorry meant the type `UserMetaData`.

Eeyore6 replied:
Now I get your point XD, my bad.

mursal-ashraf replied:
the `userTypes` file is empty

Eeyore6 replied:
Now should be fine

mursal-ashraf reviewed (commented):
Have you tried changing/updating details?

mursal-ashraf reviewed (changes requested):
Few comments/questions on the Profile page.

Eeyore6 committed:
Request update

Add a single file includes the InnerProfile, add a fail state when details not saved, remove some unnecessary statement and attributes.

Eeyore6 committed:
userTypes(single file), defaultFormData

Remove the single file for innerprofile, add a single file for userTypes(user_meta_data), add user_meta_data into the defaultFormData function.

mursal-ashraf commented on frontend/src/components/Profile/Profile.tsx:
```diff
@@ -32,122 +45,162 @@ const InnerProfile: React.FC = () => {
     setOpen(false);
   };
 
+  useEffect(() => {
```
We don't need this `useEffect` anymore after the new changes right?

Eeyore6 replied:
As we can edit, I think we need the useEffect

mursal-ashraf replied:
This one will only run when `user` changes, which is handled whenever a page rerenders. Safe to delete or not? @Imeth97 thoughts?

Eeyore6 replied:
I tried to run without the useEffect, it will not show the saved data once u refresh

Eeyore6 replied:
If you remove it, the formData will be set initially and won't be updated if the user data changes.

mursal-ashraf replied:
Yep - got it. In that case can you refactor it like this then:

```useEffect(() => {
    if (user) {
      setFormData((previousFormData) => ({
        ...previousFormData,
        ...(!!user.user_metadata && user.user_metadata),
        ...(!!user.email && { email: user.email }),
      }));
    }
  }, [user]);```

mursal-ashraf commented on frontend/src/components/Profile/Profile.tsx:
```diff
@@ -32,122 +45,162 @@ const InnerProfile: React.FC = () => {
+  }, [user]);
+
+  const handleSave = async () => {
+    try {
```
What's the try/catch for? That should be done internally by the supabase function right?

Eeyore6 replied:
I changed to "const { data, error } = await client.auth.updateUser({ data: formData });"

mursal-ashraf commented:
Can you also run `npm run format`, keeps the files neat

Eeyore6 committed:
Small Change

run 'npm run dev', update userTypes
Use internall supabase function instead of try/catch

Eeyore6 committed:
Update Profile.tsx

Update Profile.tsx

Eeyore6 committed:
Update Profile.tsx

Change the save button size/position

mursal-ashraf reviewed (approved):
One minor refactor in Profile.tsx and then good to go. Awesome work mate.

Eeyore6 committed:
Update Profile.tsx

A small refactor and move the save button next to the cancel button
